MAINT: Move transferdata into buffer-wise struct

seberg opened the pull request:
This hopefully slightly simplifies the NpyIter struct in order
to be able to add a new "Context" more easily.

---

Splitting it out for easier discussion.  I need to add a second piece of info if we go ahead with signatures proposed in NEP 43. Adding two more fields (write and read) and macros for both seemed not great either, so I thought I would propose consolidating it into a struct instead.
I first wanted to consolidate even more, but at least `strides` and `ptr` are directly exposed as arrays of intp, probably leaving only `buffers` (the base pointer holding the memory). So instead, just consolidated all the "transferinfo" needed to do the actual casting.

seberg committed:
MAINT: Move transferdata into buffer-wise struct

This hopefully slightly simplifies the NpyIter struct in order
to be able to add a new "Context" more easily.

seberg committed:
MAINT: Do not use `stransfer` to indicate NPY_OP_ITFLAG_USINGBUFFER

seberg commented on numpy/core/src/multiarray/nditer_api.c:
```diff
@@ -2443,7 +2429,12 @@ npyiter_copy_to_buffers(NpyIter *iter, char **prev_dataptrs)
+         * If OP_ITFLAG_USINGBUFFER is enabled and the read func is not NULL,
+         * the buffer needs to be read.
+         */
+        if (op_itflags[iop] & NPY_OP_ITFLAG_USINGBUFFER &&
```
I felt using the flag is easier to understand than setting `stransfer` to NULL everywhere.

seberg commented:
Does anyone know what these warnings are about, that show up in the travis build?:
```
 numpy/core/src/umath/scalarmath.c.src:555:1: note: the ABI of passing aggregates with 16-byte alignment has changed in GCC 5
```
I don't really see that they can be related to these changes.

seberg committed:
MAINT: Add variable length field at end of struct.

This (unused) field should make absolutely sure that what follows
will be intp aligned, so it might be good (although I have no real
doubts that it is already).

seberg commented:
Hmm, the last commit seems to have removed that issue. I have honestly no clue if it was related or not and would be happy to back it out to see if anyone is curious.

mattip reviewed (approved):
LGTM. Maybe there is a test missing for freeing the structure, and it might be nice to add a test for the debug print.

seberg commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -3272,13 +3254,13 @@ npyiter_allocate_transfer_functions(NpyIter *iter)
-            readtransferdata[iop] = NULL;
+        if (transferinfo[iop].read.auxdata != NULL) {
+            NPY_AUXDATA_FREE(transferinfo[iop].read.auxdata);
+            transferinfo[iop].read.auxdata = NULL;
```
Yeah, debug print should get tested. Not sure this one is reasonably testable, but let me see (datetimes may just do the trick).

seberg committed:
TST: Add test for the nditer debug_print

mattip commented:
```
E           AssertionError: assert False
E            +  where False = <built-in method startswith of str object at 0xee344710>('| Iterator SizeOf: 488')
E            +    where <built-in method startswith of str object at 0xee344710> = '| Iterator SizeOf: 264'.startswith

```

seberg commented:
Went into a rabbit hole trying to add the other test, but actually found a way I think (although not the one I expected to use. The one I expected to use found an unrelated bug of course...).

Thanks, the size of `intp` and pointers isn't fixed, so that makes sense. Just deleting should solve that issue (assuming we are OK with this stdout capture; I am struggling on capsys for the other test as well, but I think I will just give up there).

seberg committed:
BUG: Add further tests and fix existing npyiter cleanup bugs

seberg commented on numpy/core/tests/test_nditer.py:
```diff
@@ -1961,10 +1963,45 @@ def test_iter_buffered_cast_structured_type():
+        simple_arr = np.array([1, 2], dtype="i")  # succeeds but needs clean up
+        with pytest.raises(ValueError):
+            nditer((simple_arr, a), ['buffered', 'refs_ok'], [intent, intent],
+                   casting='unsafe', op_dtypes=["f", sdt2])
```
I was a bit surprised, but this should cover the error cleanup path codecov complained about (we will see).  This is almost impossible to test, because it is not well defined that the error will occur at that point, to be sure, we would basically have to create a maliciously designed `dtype` from scratch.

seberg commented on numpy/core/tests/test_nditer.py:
```diff
@@ -1961,10 +1963,45 @@ def test_iter_buffered_cast_structured_type():
+    del buf, it  # Flushing only happens during deallocate right now.
+    """)
+    res = subprocess.check_output([sys.executable, "-c", code],
+                                  stderr=subprocess.STDOUT, text=True)
```
I dunno,`capsys` (https://docs.pytest.org/en/stable/capture.html#accessing-captured-output-from-a-test-function) didn't work for me here, so rather than ditch this path completely went the ugly subprocess way...

seberg commented on numpy/core/src/multiarray/nditer_pywrap.c:
```diff
@@ -1134,6 +1134,9 @@ static void
     if (self->iter) {
+        /* Store error, so that WriteUnraisable cannot clear an existing one */
+        PyObject *exc, *val, *tb;
+        PyErr_Fetch(&exc, &val, &tb);
```
Surprisingly, there was a test that actually tripped when `dealloc` accidentally cleared an existing error.  Apparently, that is important, at least if we fail during `init`.

seberg commented:
OK, hopefully CI will be happy now. I realized that I have to use `capfd` instead of `capsys` and then I can avoid those ugly subprocess calls...

seberg committed:
TST,MAINT: Avoid subprocess for the debugprint test

seberg committed:
TST: Clear out final (nil) pointers

On windows they are printed as zeros, it would have been nice to
keep it, but isn't worth the trouble

seberg committed:
TST: Split out unraisable test, add comment and skip on pypy

seberg committed:
TST: Split out failure test and ensure it requires cleanup

Adding the array is not sufficient to hit all code lines (although
most), because auxdata is NULL for most inputs.

mattip commented:
Anything else needed here?

seberg commented:
No, its done from my side, the changes in `npyiter_pywrap.c` are unrelated (they basically fix two small bugs found while adding the new test that turned out unrelated anyway) though. I hope it doesn't find more refcount bugs in the future, but we will see ;).

mattip commented:
Thanks @seberg

seberg commented:
Thanks for the relentless reviewing! :).
